chat-plugins/blackjack.js

```diff
@@ -0,0 +1,501 @@
+		if (!this.players[user]) return this.errorMessage(user, `You aren't a player in this game.`);
+		if (this.curUser !== user.userid) return this.errorMessage(user, `It's not your turn.`);
+		this.players[user].selfUhtml = 'change';
+
+		this.giveCard(user);
+	}
+	stand(user) {
+		const player = this.players[user];
+		if (!this.started) return this.errorMessage(user, `Blackjack hasn't started yet.`);
+		if (!player) return this.errorMessage(user, `You aren't a player in this game.`);
+		if (this.curUser !== user.userid) return this.errorMessage(user, `It's not your turn.`);
+		player.status = 'stand';
+		let cards = '';
+		for (let u in player.cards) cards += `[${player.cards[u]}] `;
+		this.display(Chat.html`<br /><strong>${player.name}</strong> stands with ${cards} (${player.points})`);
+		this.clear();
+
+		this.next();
+	}
+	giveCard(user) {
+		if (this.deck.length < 1) this.deck = Dex.shuffle(this.deck.slice(0));
+		const player = (user === 'dealer' ? this.dealer : this.players[user]);
+		if (!player) return; // this should never happen
+		player.cards.push(this.deck[0]);
+
+		this.deck.splice(0, 1);
+
+		if (this.deck.length === 0) {
+			this.display(`<br />${this.dealer.name} has ran out of cards in the deck; ending game...`);
+			this.clear();
+			this.end();
+			return;
+		}
+
+		player.points = this.getCardPoints(player);
+
+		if (player.cards.length < 3) return;
+
+		if (player.cards.length > 2) this.display(Chat.html`<br /><strong>${player.name}</strong> hit and received [${player.cards[player.cards.length - 1]}] (${player.points})`);
+
+		if (user === 'dealer') {
+			if (player.points > 21) {
+				let cards = '';
+				for (let u in player.cards) cards += `[${player.cards[u]}] `;
+				this.display(Chat.html`<br /><strong>${this.dealer.name}</strong> has busted with ${cards} (${player.points})`);
+				this.end();
+				return;
+			} else if (player.points === 21) {
+				this.display(`<br /><strong>${this.dealer.name}</strong> has blackjack!`);
+				this.end();
+				return;
+			}
+		}
+
+		if (player.points > 21) {
+			player.status = 'bust';
+			let cards = '';
+			for (let u in player.cards) cards += `[${player.cards[u]}] `;
+			this.display(Chat.html`<br /><strong>${player.name}</strong> has busted with ${cards} (${player.points})`);
+			this.clear();
+		}
+		if (player.points === 21) {
+			player.status = 'stand';
+			this.display(Chat.html`<br /><strong>${player.name}</strong> has blackjack!`);
+			this.clear();
+		}
+		if (user !== 'dealer') this.players[user].cards = player.cards;
+		this.next();
+	}
+	getCardPoints(player) {
+		let points = 0;
+		let aceCount = 0;
+		for (let i in player.cards) {
+			let card = toId(player.cards[i]).toUpperCase();
+			if (!isNaN(Number(card))) {
+				points += Number(card);
+			} else if (['K', 'Q', 'J'].includes(card)) {
+				points += 10;
+			} else if (card === 'A') {
+				aceCount++;
+			}
+		}
+		for (let i = aceCount; i > 0; i--) {
```

Comment:
This still doesn't work I think - the following hand would still cause unintended busts
[A] [A]  + [K] 
At the end of the deal, the total would be at 12 (11 + 1), but after you hit again, the total should be 12 (again), but with this algorithm, you'll get a [K] (10) + [A] (11 because it's still below 21), then [A] (22).
This results in a bust.

Instead if you increase it by 10 while increasing aceCount, you could do
```
while (points > 21 && aceCount) {
    points -= 10;
    aceCount--;
}
```
This way it only decreases if it's still above 21 after the "high-ace" total is calculated